dashboard/lib/pd/survey_pipeline/daily_survey_decorator.rb

```diff
@@ -23,135 +26,108 @@ class DailySurveyDecorator < SurveyPipelineWorker
+    # Create single-workshop survey summary report to send to client.
     #
-    # @param summaries [Array<Hash>] an array of summary results.
-    # @param parsed_questions [Hash] question data get from DailySurveyParser.
-    # @param parsed_submissions [Hash] submission data get from DailySurveyParser.
-    # @param current_user [User] user making survey report request.
-    # @param errors [Array] non-fatal errors encountered when calculating survey summaries.
+    # @param data [Hash] a hash contains pieces of information from previous steps in the pipeline
+    # @option data [Array<Hash>] :summaries survey result summaries
+    # @option data [Hash] :parsed_questions questions parsed from Pd::SurveyQuestion
+    # @option data [Hash] :parsed_submissions submission parsed from Pd::Workshop(Facilitator)DailySurvey
+    # @option data [User] :current_user user requesting survey report
+    # @option data [Array<String>] :errors non-fatal errors from previous steps
     #
-    # @return [Hash] data returned to client.
+    # @return [Hash] a hash report contains 4 keys.
+    #   :course_name [String]
+    #   :questions [Hash]
+    #     {context_name => {question_scope => {question_name => question_content}}}]
+    #     context_name could be Pre Workshop, Day 1, Facilitator, Post Workshop, etc.
+    #     question_scope is either :general or :facilitator
+    #   :this_workshop [Hash]
+    #     {context_name => {:response_count => Integer, question_scope => question_summary_results}}
+    #     question_summary_results is either
+    #       :general => {question_name => summary_result}}
+    #       or :facilitator => {question_name => {facilitator_name => summary_result}}
+    #   :errors [Array<String>]
     #
-    def self.decorate(summaries:, parsed_questions:, parsed_submissions:, current_user:, errors: [])
-      result = {
+    def self.decorate_single_workshop(data)
+      report = {
         course_name: nil,
         questions: {},
         this_workshop: {},
-        all_my_workshops: {},
-        facilitators: {},
-        facilitator_averages: {},
-        facilitator_response_counts: {},
-        errors: errors
+        errors: data[:errors] || []
       }
 
-      question_by_names = index_question_by_names(parsed_questions)
+      question_by_names = index_question_by_names(data[:parsed_questions])
 
       # Populate entries for result[:this_workshop] and result[:questions]
-      summaries.each do |summary|
+      data[:summaries].each do |summary|
         # Only process summarization for specific workshops and forms.
         workshop_id, form_id, = summary.values_at(:workshop_id, :form_id)
         next unless workshop_id && form_id
 
         # Check if the current user can see specific data
-        next unless data_visible_to_user?(current_user, summary)
+        next unless summary_visible_to_user?(data[:current_user], summary)
```

Comment:
Here's an idea for later - I'd be tempted to pull any knowledge of the current user out of the decorator, and split into two decorators right here, one for the above logic that applies to any user, and one for the below logic that adds summaries.

Reply:
That makes sense. I'll keep that in mind.